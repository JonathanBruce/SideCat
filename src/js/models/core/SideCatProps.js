/* eslint-disable no-use-before-define */

import isNil from 'ramda/src/isNil';
import getType from 'ramda/src/type';
import { TYPES } from 'js/constants/TYPES';

/**
 * Creates a new Type class with the ability to validate value of its type
 * @class
 * @property {string} type Name of type
 */
class Type {
	constructor(type) {
		this.type = type;
	}

	/**
	 * Get type name
	 * @returns {string} Name of type
	 */
	getType() {
		return this.type;
	}

	/**
	 * Validate the value against its type. Returns true if matched
	 * @param   {any}     value Value to be compared
	 * @returns {boolean}       Indicator if the types match
	 */
	validate(value) {
		return getType(value) === this.type;
	}

	/**
	 * Returns decorated OptionalType object to check for nil values
	 * @returns {OptionalType} OptionalType object
	 */
	get isOptional() {
		return new OptionalType(this.type);
	}
}

/**
 * Decorated OptionalType class to return true for nil values
 * @class
 * @property {string} type Name of type
 */
class OptionalType extends Type {
	constructor(type) {
		super(type);
	}

	validate(value) {
		return isNil(value)
			? true
			: Type.prototype.validate.call(this, value);
	}
}

/**
 * Return a frozen Type object
 * @param   {string} type Name of type
 * @returns {Type}        Frozen Type object from the given param
 */
function getTypeValidator(type) {
	const typeValidator = new Type(type);

	return Object.freeze(typeValidator);
}

const SideCatProps = {};

/**
 * Validate all the props and its value type
 * @param {Object}   structure     Pre-defined structure configuration
 * @param {Model}    dataStructure Instantiated model object
 * @param {string}   modelName     Name of model
 * @returns {boolean}
 */
SideCatProps.checkPropTypes = function (structure, dataStructure, modelName) {
	return Object.keys(structure).every((propKey) => {
		const validator = structure[ propKey ];
		const value = dataStructure[ propKey ];

		if (validator.validate(value)) {
			return true;
		}

		const actualType = getType(value);
		const expectedType = validator.getType();

		throw new Error(`Invalid \`${modelName}\` structure: type ${actualType} supplied to \`${propKey}\`, expected ${expectedType}.`);
	});
};

/**
 * Assign "Structure" property to the modelClass, defined by given structure
 * @param {Model}  modelClass Model Class
 * @param {Object} structure  Structure configuration
 */
SideCatProps.defineStructure = function (modelClass, structure) {
	Object.defineProperty(modelClass, 'Structure', {
		writable: false,
		value: Object.freeze(structure)
	});
};

// Define types
SideCatProps.any = () => true;
SideCatProps.array = getTypeValidator(TYPES.ARRAY);
SideCatProps.bool = getTypeValidator(TYPES.BOOL);
SideCatProps.func = getTypeValidator(TYPES.FUNC);
SideCatProps.number = getTypeValidator(TYPES.NUMBER);
SideCatProps.object = getTypeValidator(TYPES.OBJECT);
SideCatProps.string = getTypeValidator(TYPES.STRING);

export default Object.freeze(SideCatProps);