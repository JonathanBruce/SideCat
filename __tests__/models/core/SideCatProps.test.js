/* global test, expect */
import SideCatProps from 'js/models/core/SideCatProps';

test('SideCatProps should be able to validate primitive type plus object and array', () => {
	expect(SideCatProps.array.validate([])).toBe(true);
	expect(SideCatProps.bool.validate(true)).toBe(true);
	expect(SideCatProps.func.validate(() => {})).toBe(true);
	expect(SideCatProps.number.validate(10)).toBe(true);
	expect(SideCatProps.object.validate({})).toBe(true);
	expect(SideCatProps.string.validate('')).toBe(true);
});

test('SideCatProps should be able to include Optional types', () => {
	expect(SideCatProps.array.isOptional.validate(null)).toBe(true);
	expect(SideCatProps.number.isOptional.validate(50)).toBe(true);
});

test('SideCatProps.defineStructure should be able to define a read-only "Structure" to the model passed in', () => {
	class MockModel {}

	SideCatProps.defineStructure(MockModel, {
		baz: 2,
		foo: 'bar'
	});

	expect(MockModel.Structure).toEqual({
		baz: 2,
		foo: 'bar'
	});

	expect(() => {
		MockModel.Structure.foobar = 3;
	}).toThrow();

	expect(() => {
		MockModel.Structure = { something: 'else' };
	}).toThrow();
});

test('SideCatProps.checkPropTypes should be able to validate the structure given', () => {
	const data = {
		foo: '1',
		bar: 2,
		baz: false
	};
	const structure = {
		foo: SideCatProps.string,
		bar: SideCatProps.number,
		baz: SideCatProps.bool
	};
	const valid = SideCatProps.checkPropTypes(structure, data, 'data', 'JestTest');

	expect(valid).toBe(true);
});

test('SideCatProps.checkPropTypes should property throw error properly', () => {
	const data = { foo: 'bar' };
	const structure = { foo: SideCatProps.number };

	expect(() => {
		SideCatProps.checkPropTypes(structure, data, 'JestTest');
	}).toThrowErrorMatchingSnapshot();
});