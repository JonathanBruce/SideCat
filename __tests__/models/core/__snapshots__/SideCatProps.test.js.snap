// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideCatProps.checkPropTypes should property throw error properly 1`] = `"Invalid \`JestTest\` structure: type String supplied to \`foo\`, expected Number."`;
